backends/rapidpro/status.go

```diff
@@ -341,3 +350,42 @@ func writeMsgStatuses(ctx context.Context, db *sqlx.DB, spoolDir string, statuse
 		}
 	}
 }
+
+const sqlResolveStatusMsgIDs = `
+SELECT id, channel_id, external_id 
+  FROM msgs_msg 
+ WHERE (channel_id, external_id) IN (VALUES(CAST(:channel_id AS int), :external_id))`
+
+// resolveStatusMsgIDs tries to resolve msg IDs for the given statuses - if there's no matching channel/external ID pair
+// found for a status, that status will be left with a nil msg ID.
+func resolveStatusMsgIDs(ctx context.Context, db *sqlx.DB, statuses []*DBMsgStatus) error {
+	sql, params, err := dbutil.BulkSQL(db, sqlResolveStatusMsgIDs, statuses)
+	if err != nil {
+		return err
+	}
+
+	rows, err := db.QueryContext(ctx, sql, params...)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	var msgID courier.MsgID
+	var channelID courier.ChannelID
+	var externalID string
+
+	for rows.Next() {
+		if err := rows.Scan(&msgID, &channelID, &externalID); err != nil {
+			return errors.Wrap(err, "error scanning rows")
+		}
+
+		// find the status with this channel ID and external ID and update its msg ID
+		for _, s := range statuses {
```

Comment:
Can we build a map here and loop on the statuses outside the for loop for rows?

Reply:
I guess so.. I went back and forth on that because the map has to use a compound key of channel + external id.. and I persuaded myself that the complexity of creating that isn't worth the performance gains.. but lemme try it.

Reply:
Ok it isn't too bad.. updated